pkg/controller/release/release_controller.go

```diff
@@ -313,22 +315,23 @@ func (c *Controller) syncOneReleaseHandler(key string) error {
 		c.recorder,
 	)
 
-	// This is a 2-round handler: the 1st round schedules the release on a
-	// set of clusters, and the 2nd round creates associated objects and
-	// finalizes release scheduling process. This approach comes for a
-	// reason: we are eliminating a replication lag problem. While running
-	// the system in production, we observed some cases where a few
-	// sequentual updates on the same object experienced apiserver
-	// rejections due to the passed object outdate state. This happened due
-	// to the synchronisation/replication lag in etcd. This approach helps
-	// to eliminate this as a notification will be delivered once all
-	// parties are in sync.
 	if !releaseHasClusters(rel) {
-		_, err = scheduler.ChooseClusters(rel, false)
-	} else {
-		_, err = scheduler.ScheduleRelease(rel)
+		shouldForce := false
+		if err := c.chooseClusters(rel, shouldForce); err != nil {
+			return err
+		}
+		c.recorder.Eventf(
+			rel,
+			corev1.EventTypeNormal,
+			"ClustersSelected",
+			"Set clusters for %q to %v",
+			controller.MetaKey(rel),
+			rel.Annotations[shipper.ReleaseClustersAnnotation],
+		)
 	}
 
+	rel, err = scheduler.ScheduleRelease(rel.DeepCopy())
```

Comment:
Is there reason to deep copy the release here again?

Reply:
Unrelated to the previous question, now that choosing clusters no longer saves the release, what do you think about moving the `.Update()` call here to the controller instead of buried inside the scheduler, so both the happy path and the error path are handled in the same place?

Reply:
Moved, tested :-)